sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldHaveCorrectSignature.cs

```diff
@@ -44,83 +43,116 @@ public sealed class TestMethodShouldHaveCorrectSignature : SonarDiagnosticAnalyz
+        private static SignatureValidator GetValidator(IMethodSymbol method)
         {
-            return method.GetAttributes()
-                .Select(
-                    attribute =>
-                    {
-                        if (attribute.AttributeClass.Is(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute))
-                        {
-                            return KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute;
-                        }
-                        else if (attribute.AttributeClass.Is(KnownType.NUnit_Framework_TestAttribute))
-                        {
-                            return KnownType.NUnit_Framework_TestAttribute;
-                        }
-                        else if (attribute.AttributeClass.Is(KnownType.Xunit_FactAttribute))
-                        {
-                            return KnownType.Xunit_FactAttribute;
-                        }
-                        else
-                        {
-                            return null;
-                        }
-                    })
-                .FirstOrDefault();
+            // Find the first matching attribute type in the table
+            var type = method.FindTestAttribute();
+            if (type == null)
+            {
+                return NullValidator;
+            }
+
+            attributeToConstraintsMap.TryGetValue(type, out var validator);
+            return validator;
         }
 
-        private static IEnumerable<string> GetFaults(IMethodSymbol methodSymbol, KnownType knownType)
+        private static string GetFaultMessage(IMethodSymbol methodSymbol, bool publicOnly, bool allowGenerics) =>
+            GetFaultMessageParts(methodSymbol, publicOnly, allowGenerics).ToSentence();
+
+        private static IEnumerable<string> GetFaultMessageParts(IMethodSymbol methodSymbol, bool publicOnly, bool allowGenerics)
         {
-            if (methodSymbol.DeclaredAccessibility != Accessibility.Public &&
-                knownType != KnownType.Xunit_FactAttribute)
+            if (methodSymbol.DeclaredAccessibility != Accessibility.Public && publicOnly)
             {
                 yield return MakePublicMessage;
             }
 
-            if (methodSymbol.IsGenericMethod)
+            if (methodSymbol.IsGenericMethod && !allowGenerics)
             {
                 yield return MakeNotGenericMessage;
             }
 
+            // Invariant - applies to all test methods
             if (methodSymbol.IsAsync && methodSymbol.ReturnsVoid)
             {
                 yield return MakeNonAsyncOrTaskMessage;
             }
         }
+
+        /// <summary>
+        /// Validation method. Checks the supplied method and returns the error message,
+        /// or null if there is no issue.
+        /// </summary>
+        private delegate string SignatureValidator(IMethodSymbol method);
```

Comment:
Why using the old `delegate` instead of a `Func`?

Reply:
Using the delegate provides more information about what the function is for: seeing _Func<IMethodSymbol, string>_ in the code doesn't tell me anything about what the function is supposed to do. _SignatureValidator_ gives more of an idea.

Also, the code at the point of use is easier to read with the delegate
e.g.
```
var x = new Dictionary<KnownType, Func<IMethodSymbol, string>>();
// or
var x = new Dictionary<KnownType, SignatureValidator>();
```